components/lib/nearby/src/main/java/mozilla/components/lib/nearby/NearbyConnection.kt

```diff
@@ -0,0 +1,276 @@
+ *
+ */
+class NearbyConnection(
+    private val context: Context,
+    private val name: String,
+    private val authenticate: Boolean,
+    private val listener: NearbyConnectionListener
+) {
+    // Compile-time constants
+    private val TAG = "NearbyConnection"
+    private val PACKAGE_NAME = "mozilla.components.lib.nearby"
+    private val STRATEGY = Strategy.P2P_STAR
+
+    /**
+     * The state of the connection. Changes in state are communicated through
+     * [NearbyConnectionListener.updateState].
+     */
+    public sealed class ConnectionState() {
+        val name = javaClass.simpleName
+        object Isolated : ConnectionState()
+        object Advertising : ConnectionState()
+        object Discovering : ConnectionState()
+        class Authenticating(
+            // sealed classes can't be inner, so we need to pass in the connection
+            private val nearbyConnection: NearbyConnection,
+            val neighborId: String,
+            val neighborName: String,
+            val token: String
+        ) : ConnectionState() {
+            fun accept() {
+                nearbyConnection.connectionsClient.acceptConnection(neighborId, nearbyConnection.payloadCallback)
+                nearbyConnection.updateState(ConnectionState.Connecting(neighborId, neighborName))
+            }
+            fun reject() {
+                nearbyConnection.connectionsClient.rejectConnection(neighborId)
+                // This should put us back in advertising or discovering.
+                nearbyConnection.updateState(nearbyConnection.connectionState)
+            }
+        }
+        class Connecting(val neighborId: String, val neighborName: String) : ConnectionState()
+        class ReadyToSend(val neighborEndpointId: String) : ConnectionState()
+        class Sending(val neighborEndpointId: String, val payloadId: Long) : ConnectionState()
+        class Failure(val message: String) : ConnectionState()
+    }
+
+    private var connectionsClient: ConnectionsClient = Nearby.getConnectionsClient(context)
+
+    // The can be modified in both the main thread and in callbacks. Modification occurs
+    // only in updateState(), which is synchronized.
+    private lateinit var connectionState: ConnectionState
+
+    init {
+        listener.updateState(ConnectionState.Isolated)
+    }
+
+    // This method is called from both the main thread and callbacks.
+    @Synchronized private fun updateState(cs: ConnectionState) {
+        connectionState = cs
+        listener.updateState(cs)
+    }
+
+    /**
+     * Starts advertising this device. After calling this, the state will be updated to
+     * [ConnectionState.Advertising] or [ConnectionState.Failure]. If all goes well, eventually
+     * the state will be updated to [ConnectionState.Authenticating] (if [authenticate] is true)
+     * or [ConnectionState.Connecting]. A client should call either [startAdvertising] or
+     * [startDiscovering] to make a connection, not both.
```

Comment:
Yeah, I've been wondering whether the library should provide the client with a `ConnectionState` that only provides permitted operations, rather than the `NearbyConnection`. I didn't do that in the first place since it's not an architecture I've seen before, but then I'm new to Kotlin.

Also, I can't think of a reason someone would create a connection and not either advertise or discovery. Perhaps that should be specified at the start.

It would be helpful to sit down with you, @mcomella, or @pocmo sometime to figure out the best interface.